<a href="https://colab.research.google.com/github/shubhanalytics/PGPDSBA-Projects/blob/main/Proj.01%20-%20Python%20for%20Data%20Analysis/Uber%20Drive.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

UBER DRIVE ANALYSIS : The project is based on the trips made by Uber drivers. Different aspects of the trip are analyzed by using different functions in Python.


Load the necessary libraries. Import and load the dataset with a name uber.

In [1]:
import numpy as np, pandas as pd, seaborn as sns, os
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
print(os.getcwd()) # checking the current working directory and printing in the output to get confirmed

D:\Professional\Analytics\PGPDSBA-Projects\Proj.01 - Python for Data Analysis


In [3]:
# load the csv file uberdrive.csv
uber = pd.read_csv(r'D:\Professional\Analytics\PGPDSBA-Projects\Proj.01 - Python for Data Analysis\Uber_Drive.csv') 

uber.head(4) #read the file and check the first 4 rows

,START_DATE*,END_DATE*,CATEGORY*,START*,STOP*,MILES*,PURPOSE*
0,1/1/2016 21:11,1/1/2016 21:17,Business,Fort Pierce,Fort Pierce,5.1,Meal/Entertain
1,1/2/2016 1:25,1/2/2016 1:37,Business,Fort Pierce,Fort Pierce,5.0,NaN
2,1/2/2016 20:25,1/2/2016 20:38,Business,Fort Pierce,Fort Pierce,4.8,Errand/Supplies
3,1/5/2016 17:31,1/5/2016 17:45,Business,Fort Pierce,Fort Pierce,4.7,Meeting


In [4]:
uber.tail(4)   # Displaying the last 4 records of the dataset

,START_DATE*,END_DATE*,CATEGORY*,START*,STOP*,MILES*,PURPOSE*
1151,12/31/2016 13:24,12/31/2016 13:42,Business,Karachi,Unknown Location,3.9,Temporary Site
1152,12/31/2016 15:03,12/31/2016 15:38,Business,Unknown Location,Unknown Location,16.2,Meeting
1153,12/31/2016 21:32,12/31/2016 21:50,Business,Katunayake,Gampaha,6.4,Temporary Site
1154,12/31/2016 22:08,12/31/2016 23:51,Business,Gampaha,Ilukwatta,48.2,Temporary Site


In [5]:
print(uber.shape)    # showing the shape or dimension(number of rows, number of columns) of the dataset
print(uber.size)     # showing the size(rows * columns) of the dataset

(1155, 7)
8085


In [6]:
uber.info()    # displaying the information about all the variables of the dataset

<class 'pandas.DataFrame'>
RangeIndex: 1155 entries, 0 to 1154
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   START_DATE*  1155 non-null   str    
 1   END_DATE*    1155 non-null   str    
 2   CATEGORY*    1155 non-null   str    
 3   START*       1155 non-null   str    
 4   STOP*        1155 non-null   str    
 5   MILES*       1155 non-null   float64
 6   PURPOSE*     653 non-null    str    
dtypes: float64(1), str(6)
memory usage: 63.3 KB


In the above dataset info, below are the few observations:

1. There are 1155 non-null values in first 6 columns, but there are only 653 non-null values in the 'purpose' column, which means 502 values are null in 'purpose' column. We need to review and correct it.

2. Only the 'Miles' column is in float datatype, rest all columns are in str datatype, which requires our manual review and correction.


In [7]:
uber.isnull().any()    # checking the missing values in the dataset in the boolean(True or False) format

START_DATE*    False
END_DATE*      False
CATEGORY*      False
START*         False
STOP*          False
MILES*         False
PURPOSE*        True
dtype: bool

In [8]:
uber.isnull().sum()  # checking the missing values in the dataset in the integer format

START_DATE*      0
END_DATE*        0
CATEGORY*        0
START*           0
STOP*            0
MILES*           0
PURPOSE*       502
dtype: int64

In [9]:
uber.isnull().sum().sum()   # displaying the total number of missing values across all the variables

np.int64(502)

In [10]:
uber.describe()   # Summary of the original data including : Count,Mean, Std, Min, 25%,50%,75% and max

,MILES*
count,1155.000000
mean,10.566840
std,21.579106
min,0.500000
25%,2.900000
50%,6.000000
75%,10.400000
max,310.300000


In the above summary, we can see that the maximum miles driven by an uber drive is 310, and the minimum miles is 0.50

In [11]:
df = uber.copy()  # copied the original dataset(uber) to df

In [12]:
df.head(3)

,START_DATE*,END_DATE*,CATEGORY*,START*,STOP*,MILES*,PURPOSE*
0,1/1/2016 21:11,1/1/2016 21:17,Business,Fort Pierce,Fort Pierce,5.1,Meal/Entertain
1,1/2/2016 1:25,1/2/2016 1:37,Business,Fort Pierce,Fort Pierce,5.0,NaN
2,1/2/2016 20:25,1/2/2016 20:38,Business,Fort Pierce,Fort Pierce,4.8,Errand/Supplies


In [13]:
df.columns

Index(['START_DATE*', 'END_DATE*', 'CATEGORY*', 'START*', 'STOP*', 'MILES*',
       'PURPOSE*'],
      dtype='str')

In [14]:
# Renaming the columns of the dataset
df.columns = ['start_date', 'end_date', 'category', 'starting_location', 'stop_location', 'miles','purpose']  


In [15]:
df.head(4)  #display 'df' 

,start_date,end_date,category,starting_location,stop_location,miles,purpose
0,1/1/2016 21:11,1/1/2016 21:17,Business,Fort Pierce,Fort Pierce,5.1,Meal/Entertain
1,1/2/2016 1:25,1/2/2016 1:37,Business,Fort Pierce,Fort Pierce,5.0,NaN
2,1/2/2016 20:25,1/2/2016 20:38,Business,Fort Pierce,Fort Pierce,4.8,Errand/Supplies
3,1/5/2016 17:31,1/5/2016 17:45,Business,Fort Pierce,Fort Pierce,4.7,Meeting


In [16]:
df['purpose'].value_counts()  # getting the unique PURPOSE names count

purpose
Meeting            187
Meal/Entertain     160
Errand/Supplies    128
Customer Visit     101
Temporary Site      50
Between Offices     18
Moving               4
Airport/Travel       3
Charity ($)          1
Commute              1
Name: count, dtype: int64

In [17]:
df['purpose'] = df['purpose'].replace('Charity ($)', 'Charity')  # replacing the purpose name 'Charity ($)' with 'Charity'

In [18]:
df['purpose'].value_counts()

purpose
Meeting            187
Meal/Entertain     160
Errand/Supplies    128
Customer Visit     101
Temporary Site      50
Between Offices     18
Moving               4
Airport/Travel       3
Charity              1
Commute              1
Name: count, dtype: int64

In [19]:
# Capitalizing the first letter of each word in the string columns and removing leading/trailing spaces
df['starting_location'] = df['starting_location'].str.lower().str.strip()
df['stop_location'] = df['stop_location'].str.lower().str.strip()
df['category'] = df['category'].str.lower().str.strip()
df['purpose'] = df['purpose'].str.lower().str.strip()

In [20]:
df.head(3)  # displaying the first 3 records of the dataset after cleaning the data

,start_date,end_date,category,starting_location,stop_location,miles,purpose
0,1/1/2016 21:11,1/1/2016 21:17,business,fort pierce,fort pierce,5.1,meal/entertain
1,1/2/2016 1:25,1/2/2016 1:37,business,fort pierce,fort pierce,5.0,NaN
2,1/2/2016 20:25,1/2/2016 20:38,business,fort pierce,fort pierce,4.8,errand/supplies


In [21]:
# displaying total number of unique START destinations of original dataframe 'uber'

df['starting_location'].nunique()

176

In [22]:
# displaying total number of unique STOP destinations of original dataframe 'uber'

df['stop_location'].nunique()

187

In [23]:
#Let's check the rides with the starting point as 'san fransisco'

# Now, selecting the index column as 'starting_location' from the dataset and assigning it to new variable 'df_sf'

df_sf = df.set_index('starting_location')

# selecting the rows which has starting point of San Francisco using loc function and assign it to new variable
uber_trips_SF = df_sf.loc['san francisco']


# displaying all the uber trips that has starting point of San Francisco
uber_trips_SF

,start_date,end_date,category,stop_location,miles,purpose
starting_location,,,,,,
san francisco,5/9/2016 14:39,5/9/2016 15:06,business,palo alto,20.5,between offices
san francisco,6/14/2016 16:09,6/14/2016 16:39,business,emeryville,11.6,meeting
san francisco,10/19/2016 14:02,10/19/2016 14:31,business,berkeley,10.8,NaN
san francisco,11/7/2016 19:17,11/7/2016 19:57,business,berkeley,13.2,between offices
san francisco,11/8/2016 12:16,11/8/2016 12:49,business,berkeley,11.3,meeting
san francisco,11/9/2016 18:40,11/9/2016 19:17,business,oakland,12.7,customer visit
san francisco,11/10/2016 15:17,11/10/2016 15:22,business,oakland,9.9,temporary site
san francisco,11/15/2016 20:44,11/15/2016 21:00,business,berkeley,11.8,temporary site


In [24]:
uber_trips_SF.shape

(8, 6)

In [25]:
uber_trips_SF.value_counts()   # summing up all the numerical columns of the dataframe 'uber_trips_SF'

start_date        end_date          category  stop_location  miles  purpose        
5/9/2016 14:39    5/9/2016 15:06    business  palo alto      20.5   between offices    1
6/14/2016 16:09   6/14/2016 16:39   business  emeryville     11.6   meeting            1
11/7/2016 19:17   11/7/2016 19:57   business  berkeley       13.2   between offices    1
11/8/2016 12:16   11/8/2016 12:49   business  berkeley       11.3   meeting            1
11/9/2016 18:40   11/9/2016 19:17   business  oakland        12.7   customer visit     1
11/10/2016 15:17  11/10/2016 15:22  business  oakland        9.9    temporary site     1
11/15/2016 20:44  11/15/2016 21:00  business  berkeley       11.8   temporary site     1
Name: count, dtype: int64

In [26]:
# Most popular starting point in the original dataset df

print('The most popular starting point for the uber trips is:', df['starting_location'].value_counts().idxmax())
print('_____________________')
print('See the below data to verify:\n\n', df['starting_location'].value_counts().sort_values(ascending=False).head(5))


The most popular starting point for the uber trips is: cary
_____________________
See the below data to verify:

 starting_location
cary                201
unknown location    148
morrisville          85
whitebridge          68
islamabad            57
Name: count, dtype: int64


In [27]:
# most popular dropping point from the original dataset 'df'

print('The most popular dropping point for the Uber trips is:', df['stop_location'].value_counts().idxmax())
print('_________________________')
print('See the below data to verify:\n\n', df['stop_location'].value_counts().sort_values(ascending=False).head(5))


The most popular dropping point for the Uber trips is: cary
_________________________
See the below data to verify:

 stop_location
cary                203
unknown location    149
morrisville          84
whitebridge          65
islamabad            58
Name: count, dtype: int64


In [29]:
# most frequent route taken by uber drivers (using original dataset 'df')

most_frequent = (df.groupby(['starting_location', 'stop_location']).size().reset_index(name='count_of_trips').sort_values('count_of_trips', ascending=False).head())

most_frequent

,starting_location,stop_location,count_of_trips
318,unknown location,unknown location,86
203,morrisville,cary,75
42,cary,morrisville,67
36,cary,cary,53
38,cary,durham,36


In [ ]:
#%pip install sqlalchemy

In [30]:
df.query("miles > 50")

,start_date,end_date,category,starting_location,stop_location,miles,purpose
4,1/6/2016 14:42,1/6/2016 15:49,business,fort pierce,west palm beach,63.7,customer visit
232,3/17/2016 12:52,3/17/2016 15:11,business,austin,katy,136.0,customer visit
251,3/19/2016 19:33,3/19/2016 20:39,business,galveston,houston,57.0,customer visit
268,3/25/2016 13:24,3/25/2016 16:22,business,cary,latta,144.0,customer visit
269,3/25/2016 16:52,3/25/2016 22:22,business,latta,jacksonville,310.3,customer visit
270,3/25/2016 22:54,3/26/2016 1:39,business,jacksonville,kissimmee,201.0,meeting
295,4/2/2016 12:21,4/2/2016 14:47,business,kissimmee,daytona beach,77.3,customer visit
296,4/2/2016 16:57,4/2/2016 18:09,business,daytona beach,jacksonville,80.5,customer visit
297,4/2/2016 19:38,4/2/2016 22:36,business,jacksonville,ridgeland,174.2,customer visit
298,4/2/2016 23:11,4/3/2016 1:34,business,ridgeland,florence,144.0,meeting


In [31]:
#displaying all types of purposes for the trip in an array from the dataframe 'df' with no 'NA' values

df['purpose'].unique()


<StringArray>
[ 'meal/entertain',               nan, 'errand/supplies',         'meeting',
  'customer visit',  'temporary site', 'between offices',         'charity',
         'commute',          'moving',  'airport/travel']
Length: 11, dtype: str

### let's create a bar graph between miles and purpose 

In [ ]:
df.groupby(['purpose']).sum()


In [ ]:
new_dataset = {'Airport/Travel':16.5, 'Between Offices':197.0, 'Charity($)':15.1, 'Commute':180.2, 'Customer Visit':2089.5,
              'Errand/Supplies':508.0, 'Meal/Entertain':911.7, 'Meeting':2851.3, 'Moving':18.2, 'Temporary Site':523.7}


purpose = list(new_dataset.keys())
miles = list(new_dataset.values())

fig = plt.figure(figsize = (17, 10))

plt.bar(purpose, miles, color ='green', width = 0.4)
plt.title("Total/Sum miles per purpose")
plt.xlabel("purpose")
plt.ylabel("miles")

plt.show()

## Q20. Display a dataframe of Purpose and the distance travelled for that particular Purpose. (3 points)
#### Note: Use the original dataframe without dropping "NA" values

In [ ]:
# Displaying a dataframe of Purpose and the distance travelled for that particular Purpose

uber_drives.groupby(['PURPOSE*']).sum()

## Q21. Plot number of trips vs Category of trips. (3 points)
#### Note: Use the original dataframe without dropping the 'NA' values.
#### Hint : You can make a countplot or barplot.

In [ ]:
uber_drives.groupby(['CATEGORY*']).groups



## Q22. What is proportion of miles that are covered as Business trips and what is the proportion of miles that are covered as Personal trips? (3 points)

### Note:Use the original dataframe without dropping the 'NA' values. The proportion calculation  is with respect to the 'miles' variable.

#### Hint: Proportion of miles covered as business trips= (Total Miles clocked as Business Trips)/ (Total Miles)
#### Proportion of miles covered as personal trips= (Total Miles clocked as Personal Trips)/ (Total Miles)


In [ ]:
total_miles = uber_drives['MILES*'].sum()    # sum of miles = 12204.7
print('Total Miles = ',total_miles)

print('______________________')

miles_per_category = uber_drives.groupby(['CATEGORY*']).sum()

print('\nMiles per category are shown as below = \n\n', miles_per_category)

print('______________________')

#Proportion of miles covered as business trips= (Total Miles clocked as Business Trips)/ (Total Miles)

Proportion_of_miles_covered_as_business_trips = (11487.0)/(12204.7)
print('Proportion of miles covered as business trips: ',Proportion_of_miles_covered_as_business_trips)

print('______________________')

#Proportion of miles covered as personal trips= (Total Miles clocked as Personal Trips)/ (Total Miles)

Proportion_of_miles_covered_as_personal_trips = (717.7)/(12204.7)
print('Proportion of miles covered as personal trips: ',Proportion_of_miles_covered_as_personal_trips)

